### DISCUSSION TITLE
call-gSNP Splitting

### DISCUSSION THREAD

**Discussion by yashpatel6** [uclahs-cds/pipeline-call-gSNP#134] [Ideas]
To address the large scratch usage and separate the processing steps for BAMs vs variant calling, we're proposing a splitting of the current call-gSNP pipeline:

## BAM processing pipeline
  - Handles indel realignment with GATK3.7
  - BQSR
  - Reheadering
  - Contamination/QC processes
  - Parallelization per chromosome to remove requirement of deduplication (reads getting duplicated if they overlap an interval breakpoint)

## pipeline-call-gSNP
  - Handles HaplotypeCaller and variant processing
  - Parallelization per intervals

Pros:
- Currently, parallelization between HaplotypeCaller and BAM processing (merging and deduplicating) leads to /scratch usage at roughly 3x input size from just BAMs. Splitting the pipeline would reduce that to 2x.
- Separation of logic between BAM processing and variant calling
- Ease of maintenance - call-gSNP currently is very large/complex due to handling logic for several functions. Separate pipelines would simplify this and make both parts easier to maintain

Cons:
- Potential increase in runtime due to overhead of running multiple pipelines (requires benchmarking to be certain of the impact since some processes would also be removed and reduce runtime with the split)

Any other considerations?

**Comment by tnyamaguchi** [comment]
This make sense to me. As we might eventually drop the Indel Realignment process in the long run, splitting the current call-gSNP into two would make the transition easier. Also, I think the run time might be reduced in the context of DNA-metapipeline because we should be able to parallelize call-gSNP HC and other somatic variant call pipelines.

**Comment by yashpatel6** [comment]
@pboutros Any thoughts on this?

**Comment by pboutros** [comment]
1. Strongly in favour almost simply based on the ~50% peak `/scratch` usage, but the other points make sense to me as well.
2. One thing to consider relative to the point @tyamaguchi-ucla made is, what happens if we drop indel realignment, would there be any pieces of the BAM processing that we'd want to retain?  And if so, would it make sense for those to be in a new pipeline or to be re-merged?
3. Separately, consider whether we need a new pipeline, or would be better off moving BAM processing components to `align-DNA`.

But to me those points are all academic. It seems one of our two pressing issues is space on `/scratch` (the other being node stability) and so pragmatically we have to move on solutions like this sooner than later and then backpay any technical debt incurred by rushing the design a little more than we'd like.
